src/main/java/teammates/logic/core/CommentsLogic.java

```diff
@@ -103,6 +104,54 @@ public CommentAttributes getComment(Long commentId) {
         return commentsDb.getCommentsForReceiver(courseId, recipientType, receiverEmail);
     }
     
+    /**
+     * Gets comments for a particular receiver, then filters out comments that the instructor cannot see.
+     * @param courseId
+     * @param recipientType
+     * @param receiverEmail
+     * @param instructorEmail
+     * @return List of comments visible to the instructor, directed at the receiver.
+     * @throws EntityDoesNotExistException
+     */
+    public List<CommentAttributes> getCommentsForReceiverVisibleToInstructor(
+            String courseId, CommentParticipantType recipientType, String receiverEmail, String instructorEmail)
+            throws EntityDoesNotExistException {
+        verifyIsCoursePresent(courseId, "get");
+        verifyIsInstructorOfCourse(courseId, instructorEmail);
+        List<CommentAttributes> commentsFromDb =
+                commentsDb.getCommentsForReceiver(courseId, recipientType, receiverEmail);
+        Iterator<CommentAttributes> iterator = commentsFromDb.iterator();
+        List<CommentAttributes> comments = new LinkedList<CommentAttributes>();
+        HashSet<String> commentsVisitedSet = new HashSet<String>();
+        boolean canViewCommentsFromOthers =
+                canViewCommentsFromOthers(courseId, recipientType, receiverEmail, instructorEmail);
+        
+        // add in the instructor's own comments to the list
+        while (iterator.hasNext()) {
+            CommentAttributes c = iterator.next();
+            if (c.giverEmail.equals(instructorEmail)) {
+                comments.add(c);
+                preventAppendingThisCommentAgain(commentsVisitedSet, c);
+                iterator.remove();
+            } else if (!canViewCommentsFromOthers) {
+                iterator.remove();
+            }
+        }
+        
+        // add in other instructor's comments, but only if they are visible to instructors
+        removeNonVisibleCommentsForInstructor(commentsFromDb, commentsVisitedSet, comments);
+        
+        // remove comments if the receiver is now anonymous
+        iterator = comments.iterator();
+        while (iterator.hasNext()) {
+            CommentAttributes c = iterator.next();
+            if (c.recipients.isEmpty() || "Anonymous".equals(c.recipients.iterator().next())) {
```

Comment:
Perhaps you could extract `c.recipients.iterator().next()` into an explaining variable